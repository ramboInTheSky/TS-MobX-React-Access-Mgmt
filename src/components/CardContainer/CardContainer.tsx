import React from 'react'
import Loader from 'react-loader'
import VpnKeyIcon from '@material-ui/icons/VpnKey'
import MeetingRoomIcon from '@material-ui/icons/MeetingRoom'
import DeleteIcon from '@material-ui/icons/DeleteOutline'
import EditIcon from '@material-ui/icons/Edit'
import {
    Wrapper,
    Header,
    Content,
    Item,
    MenuHeader,
    HeaderButton,
    NoItemsMessage,
} from './Components'
import { variables } from '../../constants/theme'
import { Menu, MenuItem } from '@material-ui/core'
import { LockGroup } from '../../models/LockGroups'

export interface CardContainerProps {
    type: 'rooms' | 'tags'
    items?: Array<string | LockGroup>
    removeItemFn?: Function
    button?: { title: string; fn: () => void }
}

export interface CardContainerState {
    menuOpen: { [key: string]: boolean }
    isLoading: { [key: string]: boolean }
}

export class CardContainer extends React.Component<
    CardContainerProps,
    CardContainerState
> {
    constructor(props: CardContainerProps) {
        super(props)
        this.state = {
            menuOpen: {},
            isLoading: {},
        }
    }

    public initializeItems() {
        const obj = {}
        if (this.props.items) {
            this.props.items.forEach((item: any) => {
                obj[item.name || item] = false
            })
            return obj
        }
        return {}
    }

    public componentDidMount() {
        const tempObj = this.initializeItems()
        this.setState({ menuOpen: tempObj, isLoading: tempObj })
    }

    public toggleMenu = (e: any, item: string) =>
        this.setState({
            menuOpen: {
                ...this.state.menuOpen,
                [item]: !this.state.menuOpen[item],
            },
        })

    public removeItem = async (e: any, item: any) => {
        this.setState({
            isLoading: { ...this.state.isLoading, [item]: true },
            menuOpen: {
                ...this.state.menuOpen,
                [item]: !this.state.menuOpen[item],
            },
        })
        if (this.props.removeItemFn) {
            await this.props.removeItemFn(e, item)
            this.setState({
                isLoading: { ...this.state.isLoading, [item]: false },
            })
        }
    }

    public render() {
        const { items, type, button } = this.props
        const { menuOpen, isLoading } = this.state
        return items ? (
            <Wrapper>
                <Header
                    style={{
                        backgroundColor: `${
                            type === 'tags'
                                ? variables.tagColor
                                : variables.roomColor
                        }`,
                    }}
                >
                    {type.toUpperCase()}
                    {button && (
                        <HeaderButton
                            style={{ backgroundColor: 'transparent' }}
                            className={`detail-panel-manage-button-${
                                button.title
                            }`}
                            variant="contained"
                            color="default"
                            onClick={button.fn}
                        >
                            <EditIcon />
                            {/* {button.title} */}
                        </HeaderButton>
                    )}
                </Header>
                <Content>
                    {items.length ? (
                        (items as any).map((_item: any, i: number) => {
                            const item = _item.id || _item
                            const key = `${type}-anchor-${item.replace(
                                / /g,
                                '-'
                            )}`
                            return (
                                <div key={key} id={key}>
                                    {isLoading[item] ? (
                                        <Loader loaded={!isLoading[item]} />
                                    ) : (
                                        <Item key={item}>
                                            <div>
                                                {this.props.removeItemFn ? (
                                                    <span
                                                        className={`delete-icon-for-${item}`}
                                                        onClick={(e: any) =>
                                                            this.toggleMenu(
                                                                e,
                                                                item
                                                            )
                                                        }
                                                    >
                                                        <DeleteIcon />
                                                    </span>
                                                ) : (
                                                    <span />
                                                )}
                                            </div>
                                            <div>
                                                {type === 'tags' ? (
                                                    <VpnKeyIcon />
                                                ) : (
                                                    <MeetingRoomIcon />
                                                )}
                                                {_item.name || item}
                                            </div>
                                        </Item>
                                    )}

                                    <Menu
                                        anchorEl={
                                            document.querySelector(
                                                `#${key}`
                                            ) as any
                                        }
                                        aria-label="Settings"
                                        id="Settings"
                                        title="Settings"
                                        open={menuOpen[item] || false}
                                        onClose={(e: any) =>
                                            this.toggleMenu(e, item)
                                        }
                                    >
                                        <MenuHeader>
                                            Delete{' '}
                                            <span>
                                                {type === 'tags'
                                                    ? 'Tag: '
                                                    : 'Room: '}
                                                {_item.name || item}
                                            </span>
                                        </MenuHeader>
                                        <MenuItem
                                            onClick={(e: any) =>
                                                this.removeItem(e, item)
                                            }
                                        >
                                            Yes
                                        </MenuItem>
                                        <MenuItem
                                            onClick={(e: any) =>
                                                this.toggleMenu(e, item)
                                            }
                                        >
                                            Cancel
                                        </MenuItem>
                                    </Menu>
                                </div>
                            )
                        })
                    ) : (
                        <NoItemsMessage>{`No ${type} to display`}</NoItemsMessage>
                    )}
                </Content>
            </Wrapper>
        ) : (
            <div />
        )
    }
}
