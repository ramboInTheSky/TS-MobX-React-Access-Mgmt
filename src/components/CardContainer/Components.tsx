import styled from '@emotion/styled'
import { Paper, IconButton } from '@material-ui/core'

import { variables } from '../../constants/theme'

export const Wrapper: React.ComponentType<any> = styled(Paper)`
    margin-top: ${variables.gutter};
    background-color: ${variables.componentBackgroundClear} ;
`
export const Header = styled.div`
    background-color: ${variables.componentBackground};
    color: #333;
    font-size: 1.3rem;
    height: 25px;
    padding: ${variables.indentation};
    display: flex;
    justify-content: space-between;
`

export const Content: React.ComponentType<any> = styled.div`
    background-color: ${variables.componentBackgroundClear} ;
    padding: ${variables.indentation};
    min-height: 5rem;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
`

export const Item: React.ComponentType<any> = styled(Paper)`
    background-color: ${variables.componentBackground} ;
    margin: ${variables.indentation};
    display: grid;
    grid-template-rows: 32% 50%;
    height: 5rem;
    padding: ${variables.indentation};
    min-width: 10rem;
    max-width: 14rem;
    & div:nth-child(1) {
        display: flex;
        flex-direction: row-reverse;
        cursor: pointer;
        font-size: 0.4rem;
        & svg {
            opacity: 0.5;
        }
    }
    & div:nth-child(2) {
        padding: 5px;
        position: relative;
        font-size: 1.2rem;
        text-align: right;
        & svg {
            opacity: 0.2;
            top: -36px;
            left: 0;
            position: absolute;
            font-size: 6rem;
            color: #999;
        }
    }
`
export const MenuHeader = styled.div`
    width: 100%;
    background-color: ${variables.dangerColor};
    height: 2rem;
    padding: 1rem;
    color: white;
    font-weight: 700;
    font-size: 1.2rem;
    outline: none;
    & span {
        font-weight: 400;
    }
`

export const HeaderButton: React.ComponentType<any> = styled(IconButton)`
    background-color: transparent;
    line-height: inherit ;
    color: ${variables.textColor} ;
    padding: 0 ;
`

export const NoItemsMessage = styled.div`
    margin: auto;
    color: ${variables.textColor};
`
