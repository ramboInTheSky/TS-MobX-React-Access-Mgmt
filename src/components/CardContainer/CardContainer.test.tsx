import React from 'react'
import { shallow } from 'enzyme'

import { CardContainer } from './CardContainer'

describe('<CardContainer />', () => {
    let props: any
    // const Component = (CardContainer as any).wrappedComponent
    beforeEach(() => {
        props = {
            removeFn: jest.fn(),
            type: 'rooms',
            items: [
                {
                    id: '111-123-123-123',
                    name: '1.11',
                    type: 'room 11',
                    notes: 'this is just a room',
                },
                {
                    id: '222-123-123-123',
                    name: '1.12',
                    type: 'room 12',
                    notes: 'this is the best room',
                },
            ],
        }
    })
    it('renders correctly with rooms', () => {
        expect(shallow(<CardContainer {...props} />)).toMatchSnapshot()
    })

    it('renders correctly with tags', () => {
        const newProps = {
            ...props,
            type: 'tags',
            items: ['123123123', '234234234'],
        }
        expect(shallow(<CardContainer {...props} />)).toMatchSnapshot()
    })
})
