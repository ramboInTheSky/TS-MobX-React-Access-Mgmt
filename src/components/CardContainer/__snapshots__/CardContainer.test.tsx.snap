// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardContainer /> renders correctly with rooms 1`] = `
<Styled(WithStyles(ForwardRef(Paper)))>
  <Styled(div)
    style={
      Object {
        "backgroundColor": "#eff180",
      }
    }
  >
    ROOMS
  </Styled(div)>
  <Styled(div)>
    <div
      id="rooms-anchor-111-123-123-123"
      key="rooms-anchor-111-123-123-123"
    >
      <Styled(WithStyles(ForwardRef(Paper)))
        key="111-123-123-123"
      >
        <div>
          <span />
        </div>
        <div>
          <pure(MeetingRoomIcon) />
          1.11
        </div>
      </Styled(WithStyles(ForwardRef(Paper)))>
      <WithStyles(ForwardRef(Menu))
        anchorEl={null}
        aria-label="Settings"
        id="Settings"
        onClose={[Function]}
        open={false}
        title="Settings"
      >
        <Styled(div)>
          Delete
           
          <span>
            Room: 
            1.11
          </span>
        </Styled(div)>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Yes
        </WithStyles(ForwardRef(MenuItem))>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Cancel
        </WithStyles(ForwardRef(MenuItem))>
      </WithStyles(ForwardRef(Menu))>
    </div>
    <div
      id="rooms-anchor-222-123-123-123"
      key="rooms-anchor-222-123-123-123"
    >
      <Styled(WithStyles(ForwardRef(Paper)))
        key="222-123-123-123"
      >
        <div>
          <span />
        </div>
        <div>
          <pure(MeetingRoomIcon) />
          1.12
        </div>
      </Styled(WithStyles(ForwardRef(Paper)))>
      <WithStyles(ForwardRef(Menu))
        anchorEl={null}
        aria-label="Settings"
        id="Settings"
        onClose={[Function]}
        open={false}
        title="Settings"
      >
        <Styled(div)>
          Delete
           
          <span>
            Room: 
            1.12
          </span>
        </Styled(div)>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Yes
        </WithStyles(ForwardRef(MenuItem))>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Cancel
        </WithStyles(ForwardRef(MenuItem))>
      </WithStyles(ForwardRef(Menu))>
    </div>
  </Styled(div)>
</Styled(WithStyles(ForwardRef(Paper)))>
`;

exports[`<CardContainer /> renders correctly with tags 1`] = `
<Styled(WithStyles(ForwardRef(Paper)))>
  <Styled(div)
    style={
      Object {
        "backgroundColor": "#eff180",
      }
    }
  >
    ROOMS
  </Styled(div)>
  <Styled(div)>
    <div
      id="rooms-anchor-111-123-123-123"
      key="rooms-anchor-111-123-123-123"
    >
      <Styled(WithStyles(ForwardRef(Paper)))
        key="111-123-123-123"
      >
        <div>
          <span />
        </div>
        <div>
          <pure(MeetingRoomIcon) />
          1.11
        </div>
      </Styled(WithStyles(ForwardRef(Paper)))>
      <WithStyles(ForwardRef(Menu))
        anchorEl={null}
        aria-label="Settings"
        id="Settings"
        onClose={[Function]}
        open={false}
        title="Settings"
      >
        <Styled(div)>
          Delete
           
          <span>
            Room: 
            1.11
          </span>
        </Styled(div)>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Yes
        </WithStyles(ForwardRef(MenuItem))>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Cancel
        </WithStyles(ForwardRef(MenuItem))>
      </WithStyles(ForwardRef(Menu))>
    </div>
    <div
      id="rooms-anchor-222-123-123-123"
      key="rooms-anchor-222-123-123-123"
    >
      <Styled(WithStyles(ForwardRef(Paper)))
        key="222-123-123-123"
      >
        <div>
          <span />
        </div>
        <div>
          <pure(MeetingRoomIcon) />
          1.12
        </div>
      </Styled(WithStyles(ForwardRef(Paper)))>
      <WithStyles(ForwardRef(Menu))
        anchorEl={null}
        aria-label="Settings"
        id="Settings"
        onClose={[Function]}
        open={false}
        title="Settings"
      >
        <Styled(div)>
          Delete
           
          <span>
            Room: 
            1.12
          </span>
        </Styled(div)>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Yes
        </WithStyles(ForwardRef(MenuItem))>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Cancel
        </WithStyles(ForwardRef(MenuItem))>
      </WithStyles(ForwardRef(Menu))>
    </div>
  </Styled(div)>
</Styled(WithStyles(ForwardRef(Paper)))>
`;
